src/ipa_pc/data_structures.rs

```diff
@@ -1,48 +1,104 @@
+//! Data structures for the dlog commitment scheme from [BCMS20]:
+//! * [`Parameters`], [`VerifierKey`], [`CommitterKey`]
+//! * [`Commitment`], [`Randomness`], [`Proof`], [`MultiPointProof`]
+//! and the implementation of their corresponding traits from [data_structures.rs](/src/data_structures.rs).
+//! In addition scheme-specific structs [`SuccinctCheckPolynomial`] and [`VerifierState`] are
+//! introduced.
+//!
+//! [BCMS20]: https://eprint.iacr.org/2020/499
 use crate::*;
 use crate::{PCCommitterKey, PCVerifierKey, Vec};
-use algebra::{AffineCurve, Field, PrimeField, ProjectiveCurve, ToBytes, UniformRand};
-use rand::thread_rng;
-use rand_core::RngCore;
+use algebra::{Curve, PrimeField};
 use std::{
     convert::TryFrom,
     io::{Read, Write},
     vec,
 };
 
-/// `UniversalParams` are the universal parameters for the inner product arg scheme.
+// #[cfg(test)]
+use rand::thread_rng;
+// #[cfg(test)]
+// use algebra::UniformRand;
+
+/// `Parameters` are the polynomial commitment parameters for the inner product arg scheme.
+// TODO: it seems to be artificial to have a `Parameters` struct which is the same as
+// the committer key. Let us investigate if this struct is really needed.
+// (Can't we simply implement the PCParameters trait for the struct PCCommitterKey?)
 #[derive(Derivative)]
 #[derivative(
-    Default(bound = ""),
     Clone(bound = ""),
     Debug(bound = ""),
     Eq(bound = ""),
     PartialEq(bound = "")
 )]
 #[derive(CanonicalSerialize, CanonicalDeserialize)]
-pub struct UniversalParams<G: AffineCurve> {
+pub struct Parameters<G: Curve> {
     /// The key used to commit to polynomials.
-    pub comm_key: Vec<G>,
+    pub comm_key: Vec<G::AffineRep>,
 
-    /// Some group generator.
+    /// The base point that carries the inner products evaluations.
+    /// Used by the prover and verifier.
     pub h: G,
 
-    /// Some group generator specifically used for hiding.
+    /// The base point used for hiding.
     pub s: G,
 
-    /// H(comm_key, h, s, max_degree)
+    /// The finger print  H(comm_key, h, s, max_degree)
     pub hash: Vec<u8>,
 }
 
-impl<G: AffineCurve> PCUniversalParams for UniversalParams<G> {
+impl<G: Curve> PCParameters<G> for Parameters<G> {
+    type CommitterKey = CommitterKey<G>;
+    type VerifierKey = VerifierKey<G>;
+    type Error = Error;
+
     fn max_degree(&self) -> usize {
         self.comm_key.len() - 1
     }
+
     fn get_hash(&self) -> &[u8] {
         self.hash.as_slice()
     }
-    fn copy_params(&mut self, other: &Self) {
-        self.s = other.s;
-        self.h = other.h;
+
+    /// Trims the base point vector of the setup function to a custom segment size
+    fn trim(
+        &self,
+        segment_size: usize,
+    ) -> Result<(Self::CommitterKey, Self::VerifierKey), Self::Error> {
+        // Ensure that supported_degree + 1 is a power of two
+        let supported_degree = (segment_size + 1).next_power_of_two() - 1;
```

Comment:
It's a bit weird here: segment_size should be an absolute value, but you are treating it as it was a degree (adding 1 before computing the next power of two). Is it for backward compatibility reasons ?

Reply:
This part just was not touched, so yes, we can call it "backward compatibility" =) But I am not fully understand what you mean under "absolute value" in case of segment_size

Reply:
So, before, this function was taking "supported_degree" to which corresponds, actually, supported_degree + 1 generators. Now, if we are passing the segment size, I would expect to supply to this function an exact number of generators, instead you are still treating them like if they were a "degree", thus generating "degree + 1" generators.